### PR TITLE
Output all IMUs in AP_InertialSensor example script

### PR DESCRIPTION
**Pull request by AndersonRayner**
Modified the AP_InertialSensor example script to display data from all detected IMU.  Added the public function get_count to the AP_InertialSensor class (similar to the get_count function in the AP_Compass class).

### CODE REVIEW

File: libraries/AP_InertialSensor/AP_InertialSensor.h [hunk 1]
```diff
@@ -114,6 +114,9 @@ class AP_InertialSensor : AP_AccelCal_Client
     const Vector3f     &get_accel(uint8_t i) const { return _accel[i]; }
     const Vector3f     &get_accel(void) const { return get_accel(_primary_accel); }
 
+    // return the number of accel/gyro instances.  Assume n_accel = n_gyro.
+    uint8_t get_count(void) const { return _accel_count; }
```
Comment (tridge) [inline libraries/AP_InertialSensor/AP_InertialSensor.h]: we already have get_accel_count() and get_gyro_count()
A get_count() is just misleading, as we can have a different gyro count to accel count

**Comment by tridge** [comment]
if you would like to make the example code assume that gyro count == accel count and use get_accel_count() then that would be fine. We don't want that assumption in the library though.

File: libraries/AP_InertialSensor/AP_InertialSensor.h (hunk 1 above)
Comment (AndersonRayner) [inline libraries/AP_InertialSensor/AP_InertialSensor.h]: Excellent.  I was looking for those two functions and could never find them.  Good to know they're there.

**Comment by AndersonRayner** [comment]
Agree with @guludo on the way the output should be done.  I'll make the modifications and re-commit.

**Comment by AndersonRayner** [comment]
I've updated the pull request.  

The example script now makes use of the existing << get_accel_count() >> and << get_gyro_count() >> functions (thanks @tridge).  

The script outputs all the detected accel and gyro data.  The data is prefixed with either a (h) for healthy, (x) for unhealthy, or a (-) for doesn't exist (for the case where accel_count != gyro_count) (thanks @guludo for the suggestion).  

I opted away from suppressing the outputs of the unhealthy/doesn't exist sensors.  I can change this so that white space is printed in lieu of invalid numbers however I don't think this is necessary (the printing of all zeros really should give it away).

File: libraries/AP_InertialSensor/examples/INS_generic/INS_generic.cpp [hunk 2]
```diff
@@ -23,6 +23,12 @@ void setup(void)
 
     // display initial values
     display_offsets_and_scaling();
+
+    // display number of detected accels/gyros
+    hal.console->printf("\n");
+    hal.console->printf("Number of detected accels : %d\n",ins.get_accel_count());
+    hal.console->printf("Number of detected gyros  : %d\n\n",ins.get_gyro_count());
```
Comment (guludo) [inline libraries/AP_InertialSensor/examples/INS_generic/INS_generic.cpp]: We should use `printf("... %" PRIu8 "...", ...)` or at least `%u`. Also, missing space after the coma :-)

File: libraries/AP_InertialSensor/examples/INS_generic/INS_generic.cpp [hunk 3]
```diff
@@ -108,16 +118,47 @@ static void run_test()
...
+        {
+            // determine how many sensors to print for
+            if (accel_count>gyro_count)
+            {
+                ins_count = accel_count;
+            }
+            else
+            {
+                ins_count = gyro_count;
+            }
```
Comment (guludo) [inline libraries/AP_InertialSensor/examples/INS_generic/INS_generic.cpp]: `ins_count = MAX(accel_count, gyro_count)` is simpler.

File: libraries/AP_InertialSensor/examples/INS_generic/INS_generic.cpp [hunk 4]
```diff
@@ -108,16 +118,47 @@ static void run_test()
...
+
+            // loop and print each sensor
+            for (uint8_t ii = 0; ii<ins_count; ii++)
+            {
+                hal.console->printf("%u - ",ii);
```
Comment (guludo) [inline libraries/AP_InertialSensor/examples/INS_generic/INS_generic.cpp]: PRIu8?

File: libraries/AP_InertialSensor/examples/INS_generic/INS_generic.cpp (hunk 3 above)
Comment (guludo) [inline libraries/AP_InertialSensor/examples/INS_generic/INS_generic.cpp]: Also, this value can be defined outside of the while block.

File: libraries/AP_InertialSensor/examples/INS_generic/INS_generic.cpp [hunk 5]
```diff
@@ -108,16 +118,47 @@ static void run_test()
...
+
+                // print accel
+                accel = ins.get_accel(ii);
+
+                if (ins.get_accel_health(ii)) { accel_state = 'h'; } // Healthy accel
+                else if (ii+1>accel_count) { accel_state = '-'; }    // No accel present
+                else { accel_state = 'x'; }                          // Accel present but not healthy
+
+                hal.console->printf("Accel (%c) : X:%6.2f Y:%6.2f Z:%6.2f norm:%5.2f",
```
Comment (guludo) [inline libraries/AP_InertialSensor/examples/INS_generic/INS_generic.cpp]: Why "norm"?

File: libraries/AP_InertialSensor/examples/INS_generic/INS_generic.cpp [hunk 6]
```diff
@@ -108,16 +118,47 @@ static void run_test()
...
 
-        length = accel.length();
-
-		if (counter++ % 50 == 0) {
-			// display results
-			hal.console->printf("Accel X:%4.2f \t Y:%4.2f \t Z:%4.2f \t len:%4.2f \t Gyro X:%4.2f \t Y:%4.2f \t Z:%4.2f\n", 
-								  accel.x, accel.y, accel.z, length, gyro.x, gyro.y, gyro.z);
-		}
+        // print each accel/gyro result every 50 cycles
+        if (counter++ % 50 == 0)
```
Comment (guludo) [inline libraries/AP_InertialSensor/examples/INS_generic/INS_generic.cpp]: Hm, we can save an indentation level if we use an early-continue here:
```c++
if (counter++ % 50 == 0) {
    continue;
}
```

File: libraries/AP_InertialSensor/examples/INS_generic/INS_generic.cpp [hunk 7]
```diff
@@ -108,16 +118,47 @@ static void run_test()
...
+            {
+                ins_count = gyro_count;
+            }
+
+            // loop and print each sensor
+            for (uint8_t ii = 0; ii<ins_count; ii++)
+            {
+                hal.console->printf("%u - ",ii);
+
+                // print accel
```
Comment (guludo) [inline libraries/AP_InertialSensor/examples/INS_generic/INS_generic.cpp]: No need for this comment :-)

**Comment by guludo** [comment]
@AndersonRayner I added some comments on the patch.

I'd also like to point out that the code doesn't follow coding style guidelines. See this: http://ardupilot.org/dev/docs/style-guide.html

> The script outputs all the detected accel and gyro data. The data is prefixed with either a (h) for healthy, (x) for unhealthy, or a (-) for doesn't exist (for the case where accel_count != gyro_count) (thanks @guludo for the suggestion).

My suggestion was a little bit different hehe But I'm fine with the way you did it.

**Comment by lucasdemarchi** [comment]
I incorporated most of the fixes suggested by @guludo and pushed to a branch: https://github.com/lucasdemarchi/ardupilot/commits/review/pr-3817

This is not tested, however. Will do it tomorrow when I have HW available.

**Comment by lucasdemarchi** [comment]
Tested on raspilot. Applied.

**Pull request closed without merging**